Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM_2nd.cpp

```diff
@@ -0,0 +1,540 @@
+
+#include "Utils/WarpXConst.H"
+#include <AMReX_Gpu.H>
+
+using namespace amrex;
+
+#ifdef WARPX_MAG_LLG
+// update M field over one timestep
+
+void FiniteDifferenceSolver::MacroscopicEvolveM_2nd (
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& Mfield, // Mfield contains three components MultiFab
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 >& H_biasfield, // H bias
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Bfield,
+    std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Bfield_old,
+    amrex::Real const dt,
+    std::unique_ptr<MacroscopicProperties> const& macroscopic_properties) {
+
+    if (m_fdtd_algo == MaxwellSolverAlgo::Yee)
+    {
+        MacroscopicEvolveMCartesian_2nd <CartesianYeeAlgorithm> (Mfield, H_biasfield, Bfield, Bfield_old, dt, macroscopic_properties);
+    }
+    else {
+       amrex::Abort("Only yee algorithm is compatible for M updates.");
+    }
+    } // closes function EvolveM
+#endif
+#ifdef WARPX_MAG_LLG
+    template<typename T_Algo>
+    void FiniteDifferenceSolver::MacroscopicEvolveMCartesian_2nd (
+        std::array< std::unique_ptr<amrex::MultiFab>, 3 > & Mfield,
+        std::array< std::unique_ptr<amrex::MultiFab>, 3 >& H_biasfield, // H bias
+        std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Bfield,
+        std::array< std::unique_ptr<amrex::MultiFab>, 3 > const& Bfield_old,
+        amrex::Real const dt,
+        std::unique_ptr<MacroscopicProperties> const& macroscopic_properties )
+    {
+
+        // build temporary vector<multifab,3> Mfield_prev, Mfield_error, a_temp, a_temp_static, b_temp
+        std::array< std::unique_ptr<amrex::MultiFab>, 3 > Mfield_prev;
+        std::array< std::unique_ptr<amrex::MultiFab>, 3 > Mfield_error;
+        std::array< std::unique_ptr<amrex::MultiFab>, 3 > a_temp;
+        std::array< std::unique_ptr<amrex::MultiFab>, 3 > a_temp_static;
+        std::array< std::unique_ptr<amrex::MultiFab>, 3 > b_temp;
+
+        // initialize Mfield_previous
+        for (int i = 0; i < 3; i++){
+        Mfield_prev[i].reset( new MultiFab(Mfield[i]->boxArray(),Mfield[i]->DistributionMap(),3,Mfield[i]->nGrow()));
+        Mfield_error[i].reset( new MultiFab(Mfield[i]->boxArray(),Mfield[i]->DistributionMap(),3,Mfield[i]->nGrow()));
+        MultiFab::Copy(*Mfield_prev[i],*Mfield[i],0,0,3,Mfield[i]->nGrow());
+        Mfield_error[i]->setVal(0.0);
+        }
+        // initialize a_temp, b_temp
+        for (int i = 0; i < 3; i++){
+        a_temp[i].reset( new MultiFab(Mfield[i]->boxArray(),Mfield[i]->DistributionMap(),3,Mfield[i]->nGrow()));
+        a_temp_static[i].reset( new MultiFab(Mfield[i]->boxArray(),Mfield[i]->DistributionMap(),3,Mfield[i]->nGrow()));
+        b_temp[i].reset( new MultiFab(Mfield[i]->boxArray(),Mfield[i]->DistributionMap(),3,Mfield[i]->nGrow()));
+        a_temp[i]->setVal(0.0);
+        a_temp_static[i]->setVal(0.0);
+        b_temp[i]->setVal(0.0);
```

Comment:
Yes, we can. 

The zero initialization is only to make them have reasonable values. Maybe we can delete the SetVal(0.0) at line 59, 66, 67 and 68. Reva, Andy, what do you think?

Reply:
The setvals shouldn't be needed since you compute them below.  I don't know what WarpX style guide says about initializing MultiFabs that get filled in later.  I'm fine with deleting the setvals if you want.  It actually can help find uninitialized value bugs.